mist/runtime/loss.py

```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
         loss = torch.mean(loss)
         return loss
 
+#Implement clDice:
+class soft_cldice(nn.Module):
+    def __init__(self, iter_=3, smooth = 1., exclude_background=False):
+        super(soft_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.soft_skeletonize = SoftSkeletonize(num_iter=10)
+        self.exclude_background = exclude_background
+
+    def forward(self, y_true, y_pred):
+        if self.exclude_background:
+            y_true = y_true[:, 1:, :, :]
+            y_pred = y_pred[:, 1:, :, :]
+        skel_pred = self.soft_skeletonize(y_pred)
+        skel_true = self.soft_skeletonize(y_true)
+        tprec = (torch.sum(torch.multiply(skel_pred, y_true))+self.smooth)/(torch.sum(skel_pred)+self.smooth)    
+        tsens = (torch.sum(torch.multiply(skel_true, y_pred))+self.smooth)/(torch.sum(skel_true)+self.smooth)    
+        cl_dice = 1.- 2.0*(tprec*tsens)/(tprec+tsens)
+        return cl_dice
+
+
+def soft_dice(y_true, y_pred):
+    """[function to compute dice loss]
+
+    Args:
+        y_true ([float32]): [ground truth image]
+        y_pred ([float32]): [predicted image]
+
+    Returns:
+        [float32]: [loss value]
+    """
+    smooth = 1
+    intersection = torch.sum((y_true * y_pred))
+    coeff = (2. *  intersection + smooth) / (torch.sum(y_true) + torch.sum(y_pred) + smooth)
+    return (1. - coeff)
+
+
+class soft_dice_cldice(nn.Module):
+    def __init__(self, iter_=3, alpha=0.5, smooth = 1., exclude_background=False):
+        super(soft_dice_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.alpha = alpha
```

Comment:
Other loss functions like the GSL, Boundary Loss, and HD loss use an alpha parameter.

Take a look at this page:
https://mist-medical.readthedocs.io/en/latest/advanced_topics/#weighting-schedules-for-boundary-based-loss-functions

It would be nice to make this implementation compatible with the existing alpha schedule implementations.

Reply:
I'm a little confused on alpha schedule. How should I go about implementing alpha scheduling?

Reply:
It's already implemented for you. Take a look here:

https://github.com/mist-medical/MIST/blob/de47168692118afb1910cffd9c86b52d20de9763/mist/runtime/utils.py#L645

And also here:

https://github.com/mist-medical/MIST/blob/de47168692118afb1910cffd9c86b52d20de9763/mist/runtime/run.py#L105

It's not a big deal. I'd say finish this PR first. I can show you how to do this later.